cmd/snapshot_export.go

```diff
@@ -17,38 +28,172 @@ func (o *snapshotExportOutput) toJSON()  { outputJSON(o) }
+				URL:      snapshot.PresignedURL,
+				Checksum: snapshot.MD5sum,
+			}, nil)
+		}
+
+		filePath, err = downloadExportedSnapshot(filePath, args[0], snapshot.PresignedURL)
+		if err != nil {
+			return err
+		}
+
+		if err = checkExportedSnapshot(filePath, snapshot.MD5sum); err != nil {
+			return err
+		}
+
+		return nil
+	},
 }
 
-func exportSnapshot(snapshotID string) (outputter, error) {
+func exportSnapshot(snapshotID string) (*egoscale.ExportSnapshotResponse, error) {
 	id, err := egoscale.ParseUUID(snapshotID)
 	if err != nil {
 		return nil, err
 	}
 
-	res, err := asyncRequest(&egoscale.ExportSnapshot{ID: id}, fmt.Sprintf("exporting snapshot %q", id))
+	res, err := asyncRequest(&egoscale.ExportSnapshot{ID: id}, fmt.Sprintf("Exporting snapshot %q", id))
 	if err != nil {
 		return nil, err
 	}
 
-	snapshot := res.(*egoscale.ExportSnapshotResponse)
+	return res.(*egoscale.ExportSnapshotResponse), nil
+}
+
+func downloadExportedSnapshot(filePath, snapshotID, url string) (string, error) {
+	filePath = filepath.ToSlash(filePath)
+
+	if filePath[len(filePath)-1:] == "/" {
+		filePath = filePath + snapshotID
+	}
+
+	localFilePath, err := filepath.Abs(filePath)
+	if err != nil {
+		return "", err
+	}
+
+	if filepath.Ext(localFilePath) != ".qcow2" {
+		localFilePath = localFilePath + ".qcow2"
+	}
+
+	_, err = os.Stat(localFilePath)
+	if !os.IsNotExist(err) {
+		return "", fmt.Errorf("file %q: already exists", localFilePath)
+	}
+
+	if err = os.MkdirAll(filepath.Dir(localFilePath), os.ModePerm); err != nil {
+		return "", err
+	}
+
+	file, err := os.OpenFile(localFilePath, os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return "", err
+	}
+
+	req, err := http.NewRequestWithContext(gContext, "GET", url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	client := http.DefaultClient
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
 
-	return &snapshotExportOutput{
-		URL:      snapshot.PresignedURL,
-		Checksum: snapshot.MD5sum,
-	}, nil
+	defer resp.Body.Close()
+
+	size, err := strconv.Atoi(resp.Header.Get("Content-Length"))
+	if err != nil {
+		return "", err
+	}
+
+	p := mpb.NewWithContext(gContext,
+		mpb.WithWidth(64),
+		mpb.WithRefreshRate(180*time.Millisecond),
+	)
+
+	bar := p.AddBar(int64(size), mpb.BarStyle("[=>-|"),
+		mpb.BarRemoveOnComplete(),
+		mpb.PrependDecorators(
+			decor.Name("Downloading snapshot file... "),
+			decor.OnComplete(decor.CountersKibiByte("% .2f / % .2f"), "success"),
+		),
+		mpb.AppendDecorators(
+			decor.OnComplete(decor.EwmaETA(decor.ET_STYLE_GO, 90), ""),
+			decor.OnComplete(decor.Name(" ] "), ""),
+		),
+	)
+
+	proxyReader := bar.ProxyReader(resp.Body)
+	defer proxyReader.Close()
+
+	if _, err = io.Copy(file, proxyReader); err != nil {
+		return "", err
+	}
+
+	p.Wait()
+
+	if err = file.Close(); err != nil {
+		return "", err
+	}
+
+	return localFilePath, nil
+}
+
+func checkExportedSnapshot(filePath, md5sum string) error {
+	fmt.Print("Verifying downloaded file checksum... ")
```

Comment:
As mentioned earlier, no messages should be output outside of the main CLI command function: this should be moved at caller level.